src/dev/flang/ast/Call.java

```diff
@@ -2476,23 +2477,37 @@ private boolean inferGenericLambdaResult(Resolution res,
                                            List<List<Pair<SourcePosition, AbstractType>>> foundAt)
   {
     var result = false;
-    if ((formalType.isFunctionType() || formalType.isLazyType()) &&
-        formalType.generics().get(0).isGenericArgument()
-        )
+    if (formalType.isFunctionType() || formalType.isLazyType())
       {
-        var rg = formalType.generics().get(0).genericArgument();
-        var ri = rg.index();
-        if (rg.feature() == _calledFeature && foundAt.get(ri) == null)
+        var at = actualArgType(res, formalType);
+        var frmLmdRt = formalType.generics().get(0);
+        if (frmLmdRt.isGenericArgument())
           {
-            var at = actualArgType(res, formalType);
-            if (!at.containsUndefined(true))
+            var rg = frmLmdRt.genericArgument();
+            var ri = rg.index();
+            if (rg.feature() == _calledFeature && foundAt.get(ri) == null)
               {
-                var rt = al.inferLambdaResultType(res, outer, at);
-                if (rt != null)
+                if (!at.containsUndefined(true))
                   {
-                    _generics = _generics.setOrClone(ri, rt);
+                    var rt = al.inferLambdaResultType(res, outer, at);
+                    if (rt != null)
+                      {
+                        _generics = _generics.setOrClone(ri, rt);
+                      }
+                    addPair(foundAt, ri, pos, rt);
+                    result = true;
                   }
-                addPair(foundAt, ri, pos, rt);
+              }
+          }
+        else if (at.containsUndefined(false) && frmLmdRt.dependsOnGenerics())
+          {
+            var rt = al.propagateTypeAndInferResult(res, outer, at, false);
+            // NYI What if there is more than one type parameter to infer?
```

Comment:
please add NYI-category.

I think you need to restructure the code of `extractType` and `indexOfGenericArgument` such that you only have on method that combines these two, i.e., remove `indexOfGenericArgument` and instead change `extractType` to return a list of tuples consisting for indices and corresponding actual type parameters.